bigtable/internal/testproxy/proxy_test.go

```diff
@@ -0,0 +1,448 @@
+	tpa = "localhost:9990"
+	bta = "localhost:9999"
+	rk  = "row"
+)
+
+var (
+	lis    *bufconn.Listener
+	client *pb.CloudBigtableV2TestProxyClient
+)
+
+func bufDialer(context.Context, string) (net.Conn, error) {
+	return lis.Dial()
+}
+
+// helper function to populate the in-memory BT table.
+// TODO(developer): Expose this (and bttest) as suite of test tools
+func populateTable(bts *bttest.Server) error {
+	ctx := context.Background()
+
+	conn, err := grpc.Dial(bts.Addr, grpc.WithTransportCredentials(insecure.NewCredentials()), grpc.WithBlock())
+	if err != nil {
+		return fmt.Errorf("testproxy setup: can't dial inmem Bigtable server: %v", err)
+	}
+	defer conn.Close()
+
+	adminClient, err := bigtable.NewAdminClient(ctx, "client", "instance", option.WithGRPCConn(conn), option.WithGRPCDialOption(grpc.WithBlock()))
+	if err != nil {
+		return fmt.Errorf("testproxy setup: can't create AdminClient: %v", err)
+	}
+	defer adminClient.Close()
+
+	if err := adminClient.CreateTable(ctx, tbl); err != nil {
+		return fmt.Errorf("testproxy setup: can't create table: %v", err)
+	}
+
+	// Create column families
+	count := 3
+	for i := 0; i < count; i++ {
+		cfName := fmt.Sprintf("%s%d", cf, i)
+		if err := adminClient.CreateColumnFamily(ctx, tbl, cfName); err != nil {
+			return fmt.Errorf("testproxy setup: can't create column family: %s", cfName)
+		}
+	}
+
+	// Create rows
+	dataClient, err := bigtable.NewClient(ctx, "client", "instance", option.WithGRPCConn(conn), option.WithGRPCDialOption(grpc.WithBlock()))
+	if err != nil {
+		return fmt.Errorf("testproxy setup: can't create Bigtable client: %v", err)
+	}
+	defer dataClient.Close()
+
+	t := dataClient.Open(tbl)
+
+	for fc := 0; fc < count; fc++ {
+		for cc := count; cc > 0; cc-- {
+			for tc := 0; tc < count; tc++ {
+				rmw := bigtable.NewReadModifyWrite()
+				rmw.AppendValue(fmt.Sprintf("%s%d", cf, fc), fmt.Sprintf("coll%d", cc), []byte("test data"))
+
+				_, err = t.ApplyReadModifyWrite(ctx, rk, rmw)
+				if err != nil {
+					return fmt.Errorf("testproxy setup: failure populating row: %v", err)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+/*
+TestMain has three threads that it needs to start:
+1) The mocked Bigtable service (server)
+2) The NewCloudBigtableV2TestProxyClient client that sends requests to the testproxy server.
+3) The testproxy server under test
+
+	The communication sequence looks kind of like:
+
+	TestProxyClient <=> test proxy server (what we're testing) <=> mocked BT server
+*/
+func TestMain(m *testing.M) {
+	ctx := context.Background()
+
+	// 1) Start the mocked Bigtable service
+	// This requires creating a "table" in memory
+	bts, err := bttest.NewServer(bta)
+	if err != nil {
+		log.Fatalf("testproxy setup: can't create inmem Bigtable server")
+	}
+	err = populateTable(bts)
+	if err != nil {
+		log.Fatalf("testproxy setup: can't populate mock table")
+	}
+
+	// 3) Start the test proxy server
+	lis = bufconn.Listen(buf)
+	s := newProxyServer(lis)
+	go func() {
```

Comment:
How does this goroutine exit?

Reply:
I believe that it exits when the tests are done running. My thinking was that when the main routine (the test process) ends, all child goroutines (like this one) are closed.

I'm open to suggestions for a better approach!

Reply:
Yes, that will work probably 90% of the time. Note that the grpc Server has a Stop method, which will cause Serve to finally return. I'd suggest something like this:

``` golang
s := newProxyServer(lis)
defer s.Stop()
```

Stop will also close your listener, so that's another benefit.


https://pkg.go.dev/google.golang.org/grpc#Server.Stop

Reply:
Done.